### PR TITLE
GDPR-353: Improve ual report upload & ual check

### PR DESCRIPTION
**Pull request by hyderkhanmoj**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java [hunk 1]
```diff
@@ -2,111 +2,88 @@
...
-
-        return offenderUalRepository.save(offenderUalEntity);
+    private double findJaroWinklerDistanceSimilarity(String x, String y) {
+        if (x == null && y == null) {
+            return 1.0;
+        }
+        if (x == null || y == null) {
+            return 0.0;
+        }
+        return StringUtils.getJaroWinklerDistance(x, y);
```
Comment (hyderkhanmoj) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java:59]: Deprecated & will be updated in subsequent commit

Comment (hyderkhanmoj) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java:59]: Deprecated & will be updated in subsequent commit

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java [hunk 2]
```diff
@@ -2,111 +2,88 @@
...
+            .filter(ualOff -> findJaroWinklerDistanceSimilarity(ualOff.getLastName(), offender.getLastName()) >= 0.93);
     }
 
 
-    public boolean isUnlawfullyAtLarge(OffenderNumber offenderNumber) {
-        return offenderUalRepository.findOneByOffenderNo(offenderNumber.getOffenderNumber()).isPresent();
+    private double findLeventienSimilarity(String x, String y) {
+        double maxLength = Double.max(x.length(), y.length());
+        if (maxLength > 0) {
+            return (maxLength - StringUtils.getLevenshteinDistance(x, y)) / maxLength;
```
Comment (hyderkhanmoj) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java:47]: Deprecated & will be updated in subsequent commit

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/ual/OffenderUalRepository.java [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
-    Optional<OffenderUalEntity> findOneByOffenderNo(final String offenderNo);
+    Optional<OffenderUalEntity> findOneByOffenderNoIgnoreCase(final String offenderNo);
 
-    Optional<OffenderUalEntity> findOneByOffenderBookingNoAndOffenderCroPncAndFirstNamesAndLastName(final String offenderBookingNo, final String offenderCroPnc, final String firstNames, final String lastName);
+    Optional<OffenderUalEntity> findOneByOffenderBookingNoIgnoreCase(final String offenderBookingNo);
 
-    void deleteByOffenderUalIdNotIn(final List<Long> ids);
+    Optional<OffenderUalEntity> findOneByOffenderPncIgnoreCase(@Param("pnc") final String pnc);
+
+    Optional<OffenderUalEntity> findOneByOffenderCroIgnoreCase(@Param("cro") final String cro);
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/ual/OffenderUalRepository.java:20]: I don't really understand what @Param does with a query?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java [hunk 4]
```diff
@@ -61,6 +63,12 @@ public void handlePendingDeletionReferral(final OffenderPendingDeletion event) {
                         .offenderNumber(new OffenderNumber(event.getOffenderIdDisplay()))
                         .offenceCodes(event.getOffenceCodes())
                         .alertCodes(event.getAlertCodes())
+                        .bookingNos(event.getOffenderAliases().stream().flatMap(offenderAlias -> offenderAlias.getOffenderBookings().stream()).map(OffenderBooking::getBookingNo).collect(Collectors.toSet()))
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java:66]: Might be nice to move this stream into its own method to help readability?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java [hunk 5]
```diff
@@ -2,111 +2,88 @@
...
-
-        return Optional.of(updatedReportIds);
+    public Optional<OffenderUalEntity> findUalOffender(final OffenderToCheck offender) {
+        final var ualOffender = offenderUalRepository.findOneByOffenderNoIgnoreCase(offender.getOffenderNumber().getOffenderNumber())
+            .or(() -> findOffenderByBooking(offender.getBookingNos())
+                .or(() -> findOffenderByPncs(offender.getPncs())
+                    .or(() -> findOffenderByCros(offender.getCros()))));
+
+        return ualOffender
+            .filter(ualOff -> findLeventienSimilarity(ualOff.getFirstNames(), offender.getFirstNames()) >= 0.93)
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java:37]: Possibly replace this magic number?

**Review by mikehalmamoj — DISMISSED** [review]

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/ual/OffenderUalRepository.java (hunk 3 above)
Comment (hyderkhanmoj) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/ual/OffenderUalRepository.java]: This is from the reminisce of a work around for a hibernate bug from when I was using contains. You can read about it [here](https://github.com/spring-projects/spring-data-jpa/issues/2472). In any case it is no longer required so I have removed it.

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java (hunk 4 above)
Comment (hyderkhanmoj) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java]: Updated

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java (hunk 5 above)
Comment (hyderkhanmoj) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/services/ual/UalService.java]: Updated

**Review by mikehalmamoj — APPROVED** [review]

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/ual/OffenderUalRepository.java (hunk 3 above)
Comment (mikehalmamoj) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/ual/OffenderUalRepository.java]: Ha! We bumped into a very similar issue recently with that escape parameter in a generated query - [had to replace with a custom query](https://github.com/ministryofjustice/send-legal-mail-to-prisons-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/contact/ContactRepository.kt#L21).

**Pull request merged**